### PR TITLE
Add simple math functions using ocml

### PR DESCRIPTION
**Pull request by guansong**

### CODE REVIEW

File: cuda2gcn/src/math.cl [hunk 1]
```diff
@@ -0,0 +1,279 @@
+/*===--------------------------------------------------------------------------
+ *                   ROCm Device Libraries
+ *
+ * This file is distributed under the University of Illinois Open Source
+ * License. See LICENSE.TXT for details.
+ *===------------------------------------------------------------------------*/
+
+#include "ocml.h"
+
+#define ATTR __attribute__((always_inline const))
```
Comment (b-sumner) [inline cuda2gcn/src/math.cl]: The OCML functions themselves are not all const.  I'd drop the const here for now.

File: cuda2gcn/src/math.cl [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+#define FUNC1D(root, x) \
```
Comment (b-sumner) [inline cuda2gcn/src/math.cl]: I don't really see the point of adding the parameter name to the macro argument.

File: cuda2gcn/src/math.cl [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+//-------- T __nv_tgamma
+//-------- T __nv_tgammaf
+FUNC1(tgamma)
+
+//-------- T __nv_trunc
+//-------- T __nv_truncf
+FUNC1(trunc)
+
+//-------- T __nv_abs
```
Comment (b-sumner) [inline cuda2gcn/src/math.cl]: Most of the following are not math functions, and the current implementation will require the opencl library.  Is that OK?

File: cuda2gcn/src/math.cl [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+ATTR int __nv_abs(int x) { return abs(x); }
+
+//-------- T __nv_llabs
+ATTR long __nv_llabs(long x) { return abs(x); }
+
+//-------- T __nv_max
+ATTR int __nv_max(int a, int b) { return max(a,b); }
+
+//-------- T __nv_fmax
+ATTR double __nv_fmax(double a, double b) { return max(a,b); }
```
Comment (b-sumner) [inline cuda2gcn/src/math.cl]: OCML defines fmax and fmin operations on half, float, and double types.

File: cuda2gcn/src/math.cl [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+ATTR float __nv_fmaxf(float a, float b) { return max(a,b); }
+
+//-------- T __nv_llmax
+ATTR long __nv_llmax(long a, long b) { return max(a,b); }
+
+//-------- T __nv_ullmax
+ATTR ulong __nv_ullmax(ulong a, ulong b) { return max(a,b); }
+
+//-------- T __nv_umax
+ATTR uint __nv_umax(uint a, uint b) { return max(a,b); }
```
Comment (b-sumner) [inline cuda2gcn/src/math.cl]: Consider using a < b ? b : a here and for the other integer functions.  The compiler will do the right thing.

**Comment by guansong** [comment]
Change as suggested. Thanks.

File: cuda2gcn/src/generic.cl [hunk 6]
```diff
@@ -0,0 +1,54 @@
+/*===--------------------------------------------------------------------------
+ *                   ROCm Device Libraries
+ *
+ * This file is distributed under the University of Illinois Open Source
+ * License. See LICENSE.TXT for details.
+ *===------------------------------------------------------------------------*/
+
+#define ATTR __attribute__((always_inline const))
```
Comment (b-sumner) [inline cuda2gcn/src/generic.cl]: missing comma, this can't compile

File: cuda2gcn/src/generic.cl [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ATTR ulong __nv_ullmax(ulong a, ulong b) { return MAX(a,b); }
+
+//-------- T __nv_umax
+ATTR uint __nv_umax(uint a, uint b) { return MAX(a,b); }
+
+//-------- T __nv_min
+ATTR int __nv_min(int a, int b) { return MIN(a,b); }
+
+//-------- T __nv_fmin
+ATTR double __nv_fmin(double a, double b) { return MIN(a,b); }
```
Comment (b-sumner) [inline cuda2gcn/src/generic.cl]: fmin/fmax should probably be moved back into math.cl where you can use the macros.  OCML supports fmin/fmax on all floating point types.

File: cuda2gcn/src/generic.cl (hunk 6 above)
Comment (guansong) [inline cuda2gcn/src/generic.cl]: Strang, it compiles on my machine. But will do.

File: cuda2gcn/src/generic.cl (hunk 7 above)
Comment (guansong) [inline cuda2gcn/src/generic.cl]: Right. Will do.

**Comment by guansong** [comment]
Change as suggested.

File: cuda2gcn/src/math.cl [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+//-------- T __nv_fmodf
+FUNC2(fmod)
+
+//-------- T __nv_hypot
+//-------- T __nv_hypotf
+FUNC2(hypot)
+
+//-------- T __nv_ilogb
+//-------- T __nv_ilogbf
+FUNC1(ilogb)
```
Comment (b-sumner) [inline cuda2gcn/src/math.cl]: ilogb returns int

File: cuda2gcn/src/math.cl [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+//-------- T __nv_ldexp
+//-------- T __nv_ldexpf
+FUNC2(ldexp)
```
Comment (b-sumner) [inline cuda2gcn/src/math.cl]: 2nd argument to ldexp is int

**Comment by guansong** [comment]
Good catch! Remove them to define later.

**Pull request merged**